Adapt XLA changes to develop-upstream

sumesh13 opened the pull request:
1. Port changes from fix-xla-tf-cnn-benchmark-backup
2.  Separate convolution algorithm picker logic for miopen
3. Disable code object v3 for amdgpu 
4. Changes to adapt to new changes in develop-upstream like name changes of passes, interface changes etc

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

whchung commented on tensorflow/compiler/xla/debug_options_flags.cc:
```diff
@@ -31,7 +31,9 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
   opts.set_xla_llvm_disable_expensive_passes(false);
   opts.set_xla_backend_optimization_level(3);
   opts.set_xla_cpu_multi_thread_eigen(true);
+#ifdef GOOGLE_CUDA
```
`GOOGLE_CUDA` won't be defined for XLA. Is there any better way?

whchung commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -1006,7 +1035,7 @@ cc_library(
     name = "cudnn_fused_conv_rewriter_test",
     srcs = ["cudnn_fused_conv_rewriter_test.cc"],
-    tags = tf_cuda_tests_tags(),
+    tags = if_cuda_is_configured(if_cuda(["tf_cuda_tests_tags()"])),
```
probably `if_cuda_is_configured` is enough

whchung commented on tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc:
```diff
@@ -53,9 +52,10 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/gpu/multi_output_fusion.h"
-#include "tensorflow/compiler/xla/service/gpu/pad_for_tensor_cores.h"
-#include "tensorflow/compiler/xla/service/gpu/pad_insertion.h"
+#include "tensorflow/compiler/xla/service/gpu/cudnn_conv_pad_for_tensor_cores.h"
```
`git clang-format` is required to sort header files

whchung commented on tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc:
```diff
@@ -211,11 +213,17 @@ Status OptimizeHloModule(HloModule* hlo_module, se::StreamExecutor* stream_exec,
-    pipeline.AddPass<PadInsertion>();
-    // CudnnConvolutionRewriter, PadInsertion and PadForTensorCores may add
-    // instructions which can be simplified by constant folding.
+    pipeline.AddPass<CudnnConvRewriter>();
```
@sumesh13 could you clarify do we really need these CUDNN-specific optimization passes on ROCm path?

whchung commented on tensorflow/compiler/xla/service/gpu/amdgpu_executable.cc:
```diff
@@ -41,7 +41,7 @@ AMDGPUExecutable::AMDGPUExecutable(
     int isa_version,
     std::unique_ptr<const ThunkSchedule> thunk_schedule,
-    std::unique_ptr<const HloModule> hlo_module,
+    std::unique_ptr< HloModule> hlo_module,
```
nit: `git clang-format`

whchung commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc:
```diff
@@ -334,10 +334,15 @@ StatusOr<CudnnConvParams> GetCudnnConvParams(
   params.window = &conv->window();
   params.dnums = &conv->convolution_dimension_numbers();
   params.feature_group_count = conv->feature_group_count();
+#ifdef GOOGLE_CUDA
```
`GOOGLE_CUDA` won't be defined in XLA. can we switch to check only with `TENSORFLOW_USE_ROCM`?

whchung commented on tensorflow/compiler/xla/service/gpu/gpu_layout_assignment.cc:
```diff
@@ -58,8 +58,13 @@ HeuristicLayoutAssignment(const HloInstruction* instr,
+#ifdef TENSORFLOW_USE_ROCM
+
+  if (!(instr->operand(0)->shape().element_type() == xla::PrimitiveType::F16)) {
+#elif
```
it's unclear why this check should be changed. `IsVoltaOrLater` should always evaluate to `false` on ROCm. do we want `kAllNCHW` be always taken?

sumesh13 replied:
I need to revisit ..might be IsVoltaOrLater was pulling in new dependencies. Let me confirm

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc:
```diff
@@ -157,9 +153,10 @@ std::unique_ptr<llvm::TargetMachine> GetTargetMachine(
     default:
       codegen_opt_level = CodeGenOpt::None;
   }
+  const std::string FeaturesStr = "-code-object-v3";
```
need comment to describe code object v3 is disabled

whchung commented on tensorflow/compiler/xla/service/gpu/miopen_conv_algorithm_picker.cc:
```diff
@@ -0,0 +1,339 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
@sumesh13 looking at the source code of this new optimization pass... could you elaborate the difference between its `cudnn` counterpart? I could imagine we'll be challenged with the same question when we upstream this pass.

sumesh13 replied:
Will add comment in the file

whchung commented on tensorflow/compiler/xla/service/gpu/miopen_conv_algorithm_picker.h:
```diff
@@ -0,0 +1,70 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
same here.

@sumesh13 looking at the source code of this new optimization pass... could you elaborate the difference between its `cudnn` counterpart? I could imagine we'll be challenged with the same question when we upstream this pass.

sumesh13 replied:
Will add comment in the file

whchung commented on tensorflow/compiler/xla/service/gpu/stream_executor_util.cc:
```diff
@@ -26,13 +26,15 @@ using se::dnn::DataLayout;
 using se::dnn::DataLayoutString;
 using se::dnn::FilterLayout;
 using se::dnn::FilterLayoutString;
+#if GOOGLE_CUDA
```
we want `IsVoltaOrLater`. this function should not be removed

whchung commented on tensorflow/compiler/xla/service/platform_util.cc:
```diff
@@ -53,7 +53,11 @@ string CanonicalPlatformName(const string& name) {
   }
   // "gpu" and "cuda" mean the same thing.
   if (platform_str == "gpu") {
+#ifdef TENSORFLOW_USE_ROCM
```
nit: `#if TENSORFLOW_USE_ROCM`

whchung commented on tensorflow/contrib/rccl/python/ops/rccl_ops.py:
```diff
@@ -270,8 +270,9 @@ def _maybe_load_rccl_ops_so():
-          resource_loader.get_path_to_datafile('_rccl_ops.so'))
+      library_filename = os.path.join(resource_loader.get_data_files_path(),
+                                    '_rccl_ops.so')
+      _rccl_ops_so = load_library.load_op_library(library_filename)
```
@sumesh13  could you clarify if changes in this file really needed to get XLA working? they don't seem to be relevant for the purpose of this PR

sumesh13 replied:
No, likely rebase will help

whchung commented on tensorflow/core/BUILD:
```diff
@@ -2772,6 +2772,7 @@ cc_library(
     visibility = ["//visibility:public"],
     deps = [
         "//tensorflow/core/platform/default/build_config:stream_executor_no_cuda",
+        "//tensorflow/stream_executor:rocm_platform"
```
a comment here is needed. this dependency would likely need to be removed, and it's `stream_executor_no_cuda` to be amended.

whchung commented on tensorflow/python/distribute/all_reduce.py:
```diff
@@ -25,7 +25,11 @@
+if pywrap_tensorflow.IsBuiltWithROCm():
+  from tensorflow.contrib import rccl as nccl_ops
+else:
+  from tensorflow.python.ops import nccl_ops
```
logic here doesn't seem to be related to this PR, and they are already present in `develop-upstream`. please do a rebase

whchung commented:
@sumesh13 to my understanding there are still some failed unit tests in XLA, could you help enumerate them here?

sumesh13 commented:
> @sumesh13 to my understanding there are still some failed unit tests in XLA, could you help enumerate them here?

 I see at least  3 different issues in xla/tests - 1. Some library not found. I put a fix towards that - will need to retry all the fails.  2. Symbol not resolved (will need to recheck if your fix in HIP is in 3. Timeouts (will need to retry those)

sumesh13 commented:
> @sumesh13 to my understanding there are still some failed unit tests in XLA, could you help enumerate them here?

I have addressed all the comments. I see the following issues at this point 
1. I still see the constant loading issues. I think I have the changes from the HIP branch where you put the fix. 
2. Some cpu test fails 
3. Looking for complex gemm and failing 

Besides these, on the refactoring there are some hardcoded reference to shuffle and some hardcodings in elemental_ir_emitter.cc

whchung commented:
jenkins: ok to test

whchung commented:
jenkins: retest this

whchung commented:
jenkins: retest this please

whchung commented:
need to say "please" to get the build bot going...

whchung commented:
@sumesh13 Could you let me know the test results with `//tensorflow/compiler/...` ? The test on ROCm path shows changes in this PR won't bring adversely effects to non-XLA path.

whchung commented on tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h:
```diff
@@ -71,7 +71,7 @@ class IrBuilderMixin {
   template <class... Args>
   llvm::Value* BitCast(Args&&... args) {
-    return mixin_builder()->CreaterBitCast(std::forward<Args>(args)...);
+    return mixin_builder()->CreateBitCast(std::forward<Args>(args)...);
```
it's better to reuse the previous commit and update the history via `git push --force`.

whchung commented:
@sumesh13  I'm reviewing the current test results published by you. It seems you haven't really applied my fix in HIP in https://github.com/ROCm-Developer-Tools/HIP/pull/853

For example,
```
//tensorflow/compiler/xla/tests:scalar_computations_test_gpu FAILED in 18 out of 32 in 2.6s
```
This test can actually pass with fixed HIP:
```
[----------] Global test environment tear-down
[==========] 179 tests from 2 test suites ran. (19311 ms total)
[  PASSED  ] 179 tests.
```

whchung commented:
@sumesh13 please update HIP. and provide updated test results with sharding disabled to avoid running into potential limitations from ROCR runtime:

```
bazel test --config=rocm --test_sharding_strategy=disabled //tensorflow/compiler/...
```

whchung commented on tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h:
```diff
@@ -71,7 +71,7 @@ class IrBuilderMixin {
   template <class... Args>
   llvm::Value* BitCast(Args&&... args) {
-    return mixin_builder()->CreatePointerBitCastOrAddrSpaceCast(std::forward<Args>(args)...);
+    return mixin_builder()->CreaterBitCast(std::forward<Args>(args)...);
```
@sumesh13 could you provide rationale for this commit? Are there specific test cases fixed by this?

Upstream code is actually incorrect for AMD GPU.

On AMD GPU, stack variables are by default allocated on `addrspace(5)`, not on default addrspace 0. Therefore I deliberately changed quite a few places in this commit:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/commit/3c44312177bbbab3642f145afd96cb7e5e75831a

to cope with the necessity that we sometimes need to do bitcast, and sometimes addrspacecast.

sumesh13 replied:
The upstream code is likely what needs to check if its a pointer and if it correspondingly call the appropriate builder. In this case, the values are non pointes

whchung commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc:
```diff
@@ -233,12 +233,18 @@ llvm::Value* EmitFullWarpShuffleDown(llvm::Value* value, llvm::Value* offset,
+
+    llvm::Value* value_as_int = builder->CreateBitCast(value, builder->getIntNTy(bit_width));
+    llvm::Value*  result = EmitDeviceFunctionCall(
+        "__ockl_readuplane_i32",
```
@sumesh13 logic here is AMD-only. is it possible to also encapsulate this into LLVMTargetFeature or TargetIRBuilder?

sumesh13 replied:
Yeah ...will do that when I bring in my upstream changes locally
